cstore_writer.c

```diff
@@ -310,104 +319,229 @@ CStoreWriteRow(TableWriteState *writeState, Datum *columnValues, bool *columnNul
 
 
 /*
- * CStoreEndWrite finishes a cstore data load operation. If we have an unflushed
- * stripe, we flush it. Then, we sync and close the cstore data file. Last, we
- * flush the footer to a temporary file, and atomically rename this temporary
- * file to the original footer file.
+ * CStoreEndWrite finishes a cstore data load operation. If there is an unflushed
+ * stripe, it is flushed first.
  */
 void
 CStoreEndWrite(TableWriteState *writeState)
 {
-	StringInfo tableFooterFilename = NULL;
-	StringInfo tempTableFooterFileName = NULL;
-	int renameResult = 0;
 	int columnCount = writeState->tupleDescriptor->natts;
 	StripeBuffers *stripeBuffers = writeState->stripeBuffers;
+	MemoryContext oldContext = MemoryContextSwitchTo(writeState->stripeWriteContext);
 
 	if (stripeBuffers != NULL)
 	{
-		MemoryContext oldContext = MemoryContextSwitchTo(writeState->stripeWriteContext);
-
 		StripeMetadata stripeMetadata = FlushStripe(writeState);
-		MemoryContextReset(writeState->stripeWriteContext);
 
+		/* tableFooter lives in 'higher' memory context */
 		MemoryContextSwitchTo(oldContext);
 		AppendStripeMetadata(writeState->tableFooter, stripeMetadata);
+		MemoryContextSwitchTo(writeState->stripeWriteContext);
 	}
 
-	SyncAndCloseFile(writeState->tableFile);
-
-	tableFooterFilename = writeState->tableFooterFilename;
-	tempTableFooterFileName = makeStringInfo();
-	appendStringInfo(tempTableFooterFileName, "%s%s", tableFooterFilename->data,
-					 CSTORE_TEMP_FILE_SUFFIX);
-
-	CStoreWriteFooter(tempTableFooterFileName, writeState->tableFooter);
+	CStoreWriteFooter(writeState->tableFooter, writeState->relation,
+					  writeState->logging);
 
-	renameResult = rename(tempTableFooterFileName->data, tableFooterFilename->data);
-	if (renameResult != 0)
-	{
-		ereport(ERROR, (errcode_for_file_access(),
-						errmsg("could not rename file \"%s\" to \"%s\": %m",
-							   tempTableFooterFileName->data,
-							   tableFooterFilename->data)));
-	}
-
-	pfree(tempTableFooterFileName->data);
-	pfree(tempTableFooterFileName);
+	MemoryContextSwitchTo(oldContext);
 
 	MemoryContextDelete(writeState->stripeWriteContext);
 	list_free_deep(writeState->tableFooter->stripeMetadataList);
 	pfree(writeState->tableFooter);
-	pfree(writeState->tableFooterFilename->data);
-	pfree(writeState->tableFooterFilename);
 	pfree(writeState->comparisonFunctionArray);
 	FreeColumnBlockDataArray(writeState->blockDataArray, columnCount);
 	pfree(writeState);
```

Comment:
Why isn't this just memory inside the stripeWriteContext?

Reply:
stripeWriteContext is memory context per stripe. It gets reset upon each stripe flush.

writeState is per whole write operation.  It needs to survive stripeWriteContext resets.

Reply:
Ok, there probably still should be a memory context for the whole write, rather than this granular manual management. I'm ok with opening a new issue for that tho.

Reply:
introduce a context per command and clear it when  done (in  a different PR)

Reply:
logged this in #162